Technical upgrade guide

kaitlinnewson opened the pull request:
- Adds the new technical upgrade guide
- Adds links to it from the Admin Guide and Upgrading from OJS 2 to 3 Guide
- Adds a card to the main docs hub, and a card in the dev docs area

NateWr commented:
@kaitlinnewson it looks like `kaitlinnewson:upgrade-guide` is not synced with the `main` branch. I've done a copyedit and opened a PR against your branch, but it includes [lots of unrelated commits](https://github.com/kaitlinnewson/pkp-docs/pull/2). I think if you rebase your branch that will work (or if it's easier for you to create a new one I can redo my PR).

In addition to the copyedits, I also suggested a rearrangment of the cards to move thhe admin guide and upgrade guide under the dev docs hub. My suggestion is to rewrite the dev docs card to link directly to them:

![dev-docs-card](https://user-images.githubusercontent.com/2306629/134336521-feaeecb7-61fa-4cbc-99ac-f50053cbe58f.png)

If we go this route, we should also move the content under the `/dev/` directory.

Finally, here's a list of things for discussion:

- The required tools section says the release package is required, but during the step-by-step part we download it directly. Can we remove this from the required tools?
- The section on versions links to the admin guide for more information, but I couldn't find any more info there. Is there a  particular section we can link to?
- For the tip on character encodings: let's put this info in the admin guide and just link to the admin guide.
- I think we could add a "preparation" step or expand the reuqired tools section. There are two tips in the step-by-step process that we should put before people start:
  - The tip on consulting `docs/release-notes` (this tip should also recommend looking at the release notebooks)
  - A notice about how long an upgrade may take.
- There's a command in there to "Remove the original copy of the old installation." Where does this come from? We're using `/tmp` directly. Should we use a variable for the path? And should we wait until the upgrade is succcessful to recommend doing this?
- Under the testing section, I wondered if the web server tests should be part of the steps. For example, where we disable the `.htaccess` we should also add commands for these web server things.
- Should steps 9 (testing) and 10 (restore plugins) be switched?
- The date is used in a couple of backup commands. Can we turn this into a variable?
- In this command, should the `/` prefix be left out? I think these variables already include it: `$ mv "/$OJS_WEB_PATH" "/$OJS_BACKUP_PATH"`

kaitlinnewson commented:
@NateWr I've synced the branch with `pkp-docs:main`, but it looks like there are some merge conflicts showing now.

NateWr commented:
Thanks @kaitlinnewson. I've synced up my PR now. https://github.com/kaitlinnewson/pkp-docs/pull/2 It's also easier to review the copyedits in case any of the original writers aren't happy with my changes.

kaitlinnewson commented:
For tracking remaining things to review and/or change before merge:

- [ ] The required tools section says the release package is required, but during the step-by-step part we download it directly. Can we remove this from the required tools?
- [ ] The section on versions links to the admin guide for more information, but I couldn't find any more info there. Is there a particular section we can link to?
- [ ] For the tip on character encodings: let's put this info in the admin guide and just link to the admin guide.
- [ ] I think we could add a "preparation" step or expand the required tools section. There are two tips in the step-by-step process that we should put before people start:
    - The tip on consulting docs/release-notes (this tip should also recommend looking at the release notebooks)
    - A notice about how long an upgrade may take.
- [ ] There's a command in there to "Remove the original copy of the old installation." Where does this come from? We're using /tmp directly. Should we use a variable for the path? And should we wait until the upgrade is successful to recommend doing this?
- [ ] Under the testing section, I wondered if the web server tests should be part of the steps. For example, where we disable the .htaccess we should also add commands for these web server things.
- [ ] Should steps 9 (testing) and 10 (restore plugins) be switched?
- [ ] The date is used in a couple of backup commands. Can we turn this into a variable?
- [ ] In this command, should the / prefix be left out? I think these variables already include it: $ mv "/$OJS_WEB_PATH" "/$OJS_BACKUP_PATH"

marcbria commented on upgrade-guide/en/index.md:
````diff
@@ -0,0 +1,371 @@
+Set the permissions of new files as required by your server configuration.
+
+```bash
+(Debian)$ sudo chown -R nobody:www-data /var/www/html/public/ /var/www/html/cache/
````
Debian part need to be:
 (Debian)$ sudo chown -R WEB_USER:WEB_GROUP /var/www/html/public/ /var/www/html/cache/
 
And same on line  241.

marcbria commented:
Comments about comments:

> Move content into /dev if keeping on the dev docs card
Like the idea. Not now, but more grouping (rethinking categories of documentation to talk about "Learning OJS", "Running OJS"...) will probably help to make docs home more simple/clear.

> The required tools section says the release package is required, but during the step-by-step part we download it directly. Can we remove this from the required tools?

Sorry but I think it's important to explain in "requirements" that a release package will be needed because it lets people understand what kind of upgrade they will perform (no git, no docker, no vagrant...).
At same time, this is the only point where we refer the pkp download page.

So, instead of former phrasing:
"An OJS release distributed by PKP (for example, tar/zip packages)."

What about clarify that don't need to be downloaded with something like:
"An OJS release distributed by PKP (don't download yet)".

> The section on versions links to the admin guide for more information, but I couldn't find any more info there. Is there a particular section we can link to?

Yes. It still need to be done.
As far as new "versions section" includes a basic explanation, this reference could be removed till Admin guide is extended, even I feel would be better do the change now, because the "versions" section taking a lot of space in this upgrade guide

> For the tip on character encodings: let's put this info in the admin guide and just link to the admin guide.
Good. Thanks

PD: Sorry, need to leave... following this the list in a couple of hours.

marcbria commented:
Guide refers to README to find the library/plugins requirements, but they are not included in this document.
Extending README with the missing info: https://github.com/pkp/ojs/pull/3192

kaitlinnewson commented:
>     * Under the testing section, I wondered if the web server tests should be part of the steps. For example, where we disable the `.htaccess` we should also add commands for these web server things.

@NateWr could you clarify what you're thinking should be added here?

NateWr commented:
> could you clarify what you're thinking should be added here?

[These two tasks](https://github.com/pkp/pkp-docs/pull/832/files#diff-ee2e1fe79d587fb063039f82e23148c4152531159265ddbbe9c7b551e306f759R323-R325) are not really tests to be performed. They are necessary steps to end the upgrade process similar to [removing the maintenance mode](https://github.com/pkp/pkp-docs/pull/832/files#diff-ee2e1fe79d587fb063039f82e23148c4152531159265ddbbe9c7b551e306f759R278-R283).

kaitlinnewson commented:
> 14. In the jekyll compiled version, check-list items have too much margin-bottom.

@marcbria Could you file an issue for this? This is more of a docs hub issue and not related to this guide.

marcbria commented:
Sure, but if is possible, I like to find a temporary solution for the guide as far as now this is not looking like a check-list.

To be clear, the problem is the vertical space between list items (margin-top in ul and li):

![checklist](https://user-images.githubusercontent.com/2779168/135578854-12c18ac7-df3a-4086-9526-c4590977c13b.png)

Is CSS can't be touched, alternative solutions to explore are:
- Include the list in a code/monospace block.
- Keep it as checkboxes (would be an accessibility issue for a few, but right now is an usability one for everybody).
- Move this ul/li list to paragraph+fontawsome squares with inline css styles.
- ...

If CSS can be touched, just point me the CSS file that need to be changed and I can make a PR with a new css class.

BTW, it's a nuance but initially the title of this section was not "Test the Upgrade"... was a verification list.

I'm fine with new title and new sorting of list items, but I'm telling because this is related with Nate's comment telling us that DB checks were not a test (and he is right, but we hope checking db engine and charsets will help in reducing support petitions).

marcbria commented:
I made a new review of last release available and I noticed that we perform a regular upgrade here:
https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L261

But then, a few lines after, we explain how to log it:
https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L295

I think would be better move the "Log the Output" section to a tip-block just [under the "regular" upgrade comment](https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L273) (or under the success screen, as it was in the initial version) so all code boxes will be necessary actions.

Apart from this, I miss a "Congratulations! You are done" message as steep 12.
A long road like this, needs a proper closure, don't you think?

I made this first proposal, but feel free to modify (even ignore).

`
12. Celebrate!

**You are done. Congratulations!**
Now is time to spread good news to journal Editors to let them explain to the publication users about the new features and changes.
We know upgrade process is not trivial but it's something we are working hard to improve in future versions, and is not a trivial task.
Take a look to this issues if you want to help or keep informed: https://github.com/pkp/pkp-lib/projects/23
`

kaitlinnewson commented:
> I made a new review of last release available and I noticed that we perform a regular upgrade here:
> 
> https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L261
> 
> But then, a few lines after, we explain how to log it:
> 
> https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L295
> 
> I think would be better move the "Log the Output" section to a tip-block just [under the "regular" upgrade comment](https://github.com/pkp/pkp-docs/blob/b49c0a4217cf290ff782fb5992a41ff8e5063dff/dev/upgrade-guide/en/index.md#L273) (or under the success screen, as it was in the initial version) so all code boxes will be necessary actions.
> 
> Apart from this, I miss a "Congratulations! You are done" message as steep 12. A long road like this, needs a proper closure, don't you think?
> 
> I made this first proposal, but feel free to modify (even ignore).
> 
> ```
> ### 12. Celebrate!
> 
> **You are done. Congratulations!**
> Now is time to spread good news to journal Editors to let them explain to the publication users about the new features and changes.
> We know upgrade process is not trivial but it's something we are working hard to improve in future versions, and is not a trivial task.
> Take a look to this issues if you want to help or keep informed: https://github.com/pkp/pkp-lib/projects/23
> ```


@marcbria I've made a couple of updates based on this:

- [Logging the output](https://deploy-preview-1--nifty-liskov-5ad8f0.netlify.app/dev/upgrade-guide/en/#8-run-the-upgrade)
- [Celebrate step](https://deploy-preview-1--nifty-liskov-5ad8f0.netlify.app/dev/upgrade-guide/en/#12-celebrate)

marcbria commented:
About the "celebrate step" is good for me. Let's see what Nate says.
About the "Logging the output" why not adding it as a tip? In the end it is what it is... an optional tip that improves the usual way of use.
About the "temporary solution for checklist vertical margin" which [workaround](https://github.com/pkp/pkp-docs/pull/832#issuecomment-931968875) do you like most?

Thanks for your work Kaitlin!

NateWr commented:
I disagree about the logging change. I don't think we should weave that in with the commands, because it forks the step-by-step process. The commands are duplicated, and we have to add descriptions like "if you did this, then do that", which are likely to be missed.

I prefer separating it out entirely to its own section. There is no need to improve how you run the commands the first time through, so the user doesn't need to know about it up front.

> About the "Logging the output" why not adding it as a tip? In the end it is what it is... an optional tip that improves the usual way of use.

A tip needs to be very short to work from a visual/UX perspective. I moved it out of a tip because the tip was long and difficult to read in the yellow box.

> ### 12. Celebrate
> **Your upgrade is done. Congratulations!**
> The upgrade process can be complex, but the PKP team is working on improvements. See [the "Upgrades" GitHub project](https://github.com/pkp/pkp-lib/projects/23) for more information.

I'm sorry to be a grump about this, but I don't want to apologize for the upgrade process. These are normal system administration tasks required to safely conduct a backup and upgrade. We should be setting a tone that says "this is the level of competency we expect for you to run this publishing system". 

I'm not sure a celebration step makes a lot of sense here. This goes back to the oddity where testing is conducted before plugins are restored. The upgrade is complete after step 8, so that would be the place for a "You're done now" message. But of course in most cases the work is just beginning for the service provider, who must conduct tests and training.

I think the problem arises from the fact that the upgrade guide conducts the upgrade in production. For that reason, step 8 recommends restoring public access to the site, but we haven't even conducted tests on it yet. In reality, the tests should be conducted in a sandbox environment, so steps 8, 9, 10 and 11 are out of order. It should be:

- Complete upgrade in sandbox
- Run tests
- Move production into maintenance mode
- Complete upgrade in maintenance mode
- Remove maintenance mode
- Remove backups

marcbria commented:
**About "Logging the output"...**

There is no middle ground and I think we've both run out of arguments, so unless Kaitlin has a strong opinion on this and helps us break the tie, I concede and suggest doing as Nate asks... or even substitute the "simple" command (php tools/upgrade.php check) for the "extended" version (as Nate suggested in the hackMD comments).

**About "12. Celebrate"...**

> I'm sorry to be a grump about this, but I don't want to apologize for the upgrade process. 
> I'm not sure a celebration step makes a lot of sense here. 

You are absolutely right about this, there was something that didn't convince me and I couldn't tell what it was. 
While I think it's good to show that work is being and will be done to make the upgrade even easier, it's not a reason to apologize. We can remove this "I'm sorry part" part, but I still do see the need for a final step to provide closure to the whole process... that as far as this doc is written with sysOps in mind, it won't happen until the site is back in production..

Unfortunately I can't help you much in finding the right "tone". :-(

> I think the problem arises from the fact that the upgrade guide conducts the upgrade in production. 

The guide assists in the production upgrade process and step "4. Create Sandbox" is a recommendation (which I wouldn't leave home without).

The reason for this is that we didn't want to make an eternal guide, covering absolutely everything and lose the reader in the detail... but the fact is, introducing this "step 4" we generate a bifurcation in the guide since there will be those who are working in production and those who are working in sandbox.

In that sense I think you are right and we would miss a warning or a step to remember that if they are working in sandbox, they must repeat the whole process in production (ie: "11. Do the upgrade in your production site")... otherwise I'm worried we will need to rewrite a lot of stuff and we will never release this document.

BTW Kaitlin, apologies for all these late changes. If you think I can help in something, let me know.

kaitlinnewson commented:
@NateWr @marcbria I've moved the logging section back, and instead added a link to the section in the text so that users are aware of it. My concern was that they would run it without knowing about the other option, but I agree that forking the process is confusing.

I've also updated the text in the last step to remove the "apology" text and just note that they have completed the instance upgrade. I agree that it's nice to have some kind of indicator that they have finished the technical upgrade steps, even though it is optional.

At this point I would like to do a final check for any last changes to the content of this guide. If there are still major disagreements about the content, then please open a group conversation on Slack or we can schedule a short meeting to discuss. Issues related to the docs hub itself can be looked at after this has been published.

marcbria commented:
> I've moved the logging section back, and instead added a link to the section in the text so that users are aware of it. My concern was that they would run it without knowing about the other option, but I agree that forking the process is confusing.

I shared your concern. Your solution is perfect.

> I've also updated the text in the last step to remove the "apology" text and just note that they have completed the instance upgrade. I agree that it's nice to have some kind of indicator that they have finished the technical upgrade steps, even though it is optional.

Great. Thanks.

> At this point I would like to do a final check for any last changes to the content of this guide. 

My only but to close this is the [check-list rendering](https://github.com/pkp/pkp-docs/pull/832#issuecomment-931968875). Sorry to be heavy with this but just as Nate argues that TIPs/Warnings can't be too long for UI/UX reasons, this eternal listing breaks the guide.

Can we explore some of the options I discussed in the post?

1. Add a new style to the theme css.
2. Include the list in a code/monospace block.
3. Keep it as checkboxes (would be an accessibility issue for a few, but right now is an usability one for everybody).
4. Move this ul/li list to paragraph+fontawsome squares with inline css styles.
5. ...

If you tell me your preference and the files to modify, I think I will be able to generate the css or html that would be needed for a temporary or even definitive solution.

NateWr commented:
Thanks @kaitlinnewson! I agree we can publish and iterate from there.

> My only but to close this is the check-list rendering.

I'd say for now go with the code block. I agree the styles could be reviewed, but a CSS change will impact every document in the hub, so we'd need to review it carefully.

marcbria commented:
> I'd say for now go with the code block. I agree the styles could be reviewed, but a CSS change will impact every document in the hub, so we'd need to review it carefully.

I happily buy the "code block" temporary solution but to be clear, I was not talking about changing existing list style... I was suggesting adding a new style for checklists that won't collide with any former one.

kaitlinnewson commented:
@marcbria [Here's the checklist in the code block](https://deploy-preview-1--nifty-liskov-5ad8f0.netlify.app/dev/upgrade-guide/en/#10-test-the-upgrade) - is this okay as a workaround for now? If so I think we're ready to merge!
